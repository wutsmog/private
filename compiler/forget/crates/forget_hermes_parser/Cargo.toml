[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }