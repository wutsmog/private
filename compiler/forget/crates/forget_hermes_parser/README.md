# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.