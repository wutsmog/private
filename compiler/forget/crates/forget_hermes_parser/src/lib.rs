mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
