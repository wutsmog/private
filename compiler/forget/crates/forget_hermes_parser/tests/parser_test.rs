use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
