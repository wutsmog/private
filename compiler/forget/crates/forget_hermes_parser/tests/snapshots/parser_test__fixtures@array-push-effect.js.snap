---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/array-push-effect.js
---
Input:
// arrayInstance.push should have the following effects:
//  - read on all args (rest parameter)
//  - mutate on receiver
function Component(props) {
  const x = foo(props.x);
  const y = { y: props.y };
  const arr = [];
  arr.push({});
  arr.push(x, y);
  return arr;
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 131,
          "end": 140
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 141,
            "end": 146
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 158,
                    "end": 159
                  }
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "foo",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 162,
                      "end": 165
                    }
                  },
                  "arguments": [
                    {
                      "type": "MemberExpression",
                      "object": {
                        "type": "Identifier",
                        "name": "props",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 166,
                          "end": 171
                        }
                      },
                      "property": {
                        "type": "Identifier",
                        "name": "x",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 172,
                          "end": 173
                        }
                      },
                      "computed": false,
                      "loc": null,
                      "range": {
                        "start": 166,
                        "end": 173
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 162,
                    "end": 174
                  }
                },
                "loc": null,
                "range": {
                  "start": 158,
                  "end": 174
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 152,
              "end": 175
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "y",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 184,
                    "end": 185
                  }
                },
                "init": {
                  "type": "ObjectExpression",
                  "properties": [
                    {
                      "type": "Property",
                      "key": {
                        "type": "Identifier",
                        "name": "y",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 190,
                          "end": 191
                        }
                      },
                      "value": {
                        "type": "MemberExpression",
                        "object": {
                          "type": "Identifier",
                          "name": "props",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 193,
                            "end": 198
                          }
                        },
                        "property": {
                          "type": "Identifier",
                          "name": "y",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 199,
                            "end": 200
                          }
                        },
                        "computed": false,
                        "loc": null,
                        "range": {
                          "start": 193,
                          "end": 200
                        }
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": false,
                      "computed": false,
                      "loc": null,
                      "range": {
                        "start": 190,
                        "end": 200
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 188,
                    "end": 202
                  }
                },
                "loc": null,
                "range": {
                  "start": 184,
                  "end": 202
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 178,
              "end": 203
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "arr",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 212,
                    "end": 215
                  }
                },
                "init": {
                  "type": "ArrayExpression",
                  "elements": [],
                  "loc": null,
                  "range": {
                    "start": 218,
                    "end": 220
                  }
                },
                "loc": null,
                "range": {
                  "start": 212,
                  "end": 220
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 206,
              "end": 221
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "CallExpression",
              "callee": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "arr",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 224,
                    "end": 227
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "push",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 228,
                    "end": 232
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 224,
                  "end": 232
                }
              },
              "arguments": [
                {
                  "type": "ObjectExpression",
                  "properties": [],
                  "loc": null,
                  "range": {
                    "start": 233,
                    "end": 235
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 224,
                "end": 236
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 224,
              "end": 237
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "CallExpression",
              "callee": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "arr",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 240,
                    "end": 243
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "push",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 244,
                    "end": 248
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 240,
                  "end": 248
                }
              },
              "arguments": [
                {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 249,
                    "end": 250
                  }
                },
                {
                  "type": "Identifier",
                  "name": "y",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 252,
                    "end": 253
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 240,
                "end": 254
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 240,
              "end": 255
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "Identifier",
              "name": "arr",
              "typeAnnotation": null,
              "loc": null,
              "range": {
                "start": 265,
                "end": 268
              }
            },
            "loc": null,
            "range": {
              "start": 258,
              "end": 269
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 148,
          "end": 271
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 122,
        "end": 271
      },
      "loc": null,
      "range": {
        "start": 122,
        "end": 271
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 122,
    "end": 271
  }
}
